### PR TITLE
Finish Implementation of INTERACTION SUCCESS and End game callbacks/functions

### PR DESCRIPTION
**Pull request by yummyblabla**
- [x] Fix INTERACTION SUCCESS callback

- [x] Game sends WIN_ROUND appropriately

- [x] Implement WINNING_TILES callback

- [x] Render basic info for winner tiles

- [x] Implement UPDATE_GAME_STATE callback

- [x] Implement DRAW_ROUND callback

- [x] Render basic info for round draw

- [x] Implement functions that reset instance states (i.e. hand, game, wall)

- [x] Add tests for above functions

### CODE REVIEW

File: src/pages/Game.tsx [hunk 1]
```diff
@@ -411,6 +448,43 @@ const GamePage = ({ ws, currentUser }: GameProps): JSX.Element => {
...
+    const data = payload as UpdateGameStateJSON;
+    const { dealer, wind } = data;
+    const mjGameState = gameState as MahjongGameState;
+    mjGameState.gameStateSync(dealer, wind);
+    mjGameState.resetEverything();
+  };
+
+  /**
+   * For DRAW_ROUND
+   * @param payload ???????
```
Comment (jzc1828) [inline src/pages/Game.tsx:479]: I set the payload to be:
```ts
export interface DrawRoundPayload {
  gameId: string;
  connectionId: string;
}
```
in the backend. Not quite sure if it needs anything in the payload, so I included something that might be useful to the client. Let me know if I need to add anything else to this payload

File: src/modules/mahjong/MahjongPlayer/MahjongPlayer.ts [hunk 2]
```diff
@@ -361,7 +368,11 @@ class MahjongPlayer extends UserEntity {
           console.log(event.target);
           if (canPlayTile) {
             this.getHand().setCannotPlayTile();
-            // TODO: send message to backend WIN_ROUND
+            const pointsResult = PointValidator.validateHandPoints(canWin);
+            const payload = {
+              handPointResults: pointsResult.largestHand,
+            };
```
[the remark this answers is not in the document]
Comment (yummyblabla) [reply src/modules/mahjong/MahjongPlayer/MahjongPlayer.ts:374]: gameId gets added to the payload in the callback, but I forgot to add the key WIN_ROUND in the callback. Will fix.

File: src/pages/Game.tsx [hunk 3]
```diff
@@ -338,22 +343,46 @@ const GamePage = ({ ws, currentUser }: GameProps): JSX.Element => {
...
-      const { connectionId, playedTiles, meldType } = data;
+      const { connectionId, playedTiles: playedTilesStr, meldType } = data;
       // Remove last tile from deadpile
       const playedTile = mjGameState.getDeadPile().removeLastTile();
-
-      if (connectionId && playedTiles && meldType) {
-        const tiles = playedTiles.map((tileStr) => TileFactory.createTileFromStringDef(tileStr));
+      if (connectionId && playedTilesStr && meldType) {
+        const playedTiles = playedTilesStr.map((tileStr) => TileFactory.createTileFromStringDef(tileStr));
         const userIndex = mjGameState.getUsers().findIndex((user) => user.getConnectionId() === connectionId);
```
[the remark this answers is not in the document]
Comment (yummyblabla) [reply src/pages/Game.tsx:351]: Makes sense, I have to always convert string[] to Tile[]

File: src/pages/Game.tsx (hunk 1 above)
Comment (yummyblabla) [reply src/pages/Game.tsx:479]: ConnectionId, being the person who can't draw? I'm not sure what to put here as well. I'll put in those 2 as placeholders for now.

Comment (jzc1828) [reply src/pages/Game.tsx]: ConnectionId is the user who failed to draw a tile, so the user before it got the last tile from the wall.

**Pull request merged**